### DISCUSSION TITLE
Is there a recommended way to debug a tauri application (react used as frontend) in vscode on Windows?

### DISCUSSION THREAD

**Discussion by ReinhardKrisam** [tauri-apps/tauri#4210] [Q&A] [answered]
Hello everybody,

is there a specific resource that describes the setup of the debugger in vscode for an tauri app (for windows)?

I tried out the resources i found online, but it does not seem to work properly.

My steps:

1.) create the tauri app with "npx create-tauri-app"
2.) step through the init-process (and selectin react with JS as the frontend)
3.) in vscode, the "rust-analyzer"-, the "CodeLLDB"- and the "C/C++"-addons are installed
4.) allow to set breakpoints everywere (in the user-settings of vscode)
5.) creating the launch.json with some predefined content.

Now I am not sure how to continue.
I have done some copy&paste from online-resources into the lauch.json file but nothing really helped me (what a suprise...)

The tauri-application (the rust side at least) was loaded and I could set breakpoints in the "main.rs" file that actutally worked.
But then the tauri-application window (the GUI) was loaded with an "...this page is not available..." & "localhost has denied..." message.

I've attached the launch.json (as a text-file) and a screenshot of the GUI with the error-message.

When I am starting the tauri app right after the creation (npx create-tauri-app), everything works fine. Also the react part works.

Maybe someone can help me fixing this issue.

Many thanks,

Reinhard

[launch.json.txt](https://github.com/tauri-apps/tauri/files/8770711/launch.json.txt)
![guiNotLoading](https://user-images.githubusercontent.com/19156515/170253669-6eac2877-0d1e-4098-a87a-5ac35b639350.PNG)

**Comment by FabianLars** [comment, marked as answer]
Try starting the frontend yourself in a separate terminal (for example `yarn start`), because starting the exe directly and/or not using `tauri dev` doesn't automatically execute the configured `beforeDevCommand`...

Alternatively you can do `tauri build --debug` which runs the `beforeBuildCommand` and bundles the frontend into the binary (slower to compile).

**Reply by ReinhardKrisam** [reply]
Thank you for your fast answer. I will try it.

**Reply by ReinhardKrisam** [reply]
It works well to start the GUI now. Thank you.
I started the frontend in a separat terminal.


The only thing thats not working right now is the invoke of a rust function from the react side.
(e.g. a button press should invoke a rust function that does a "println")

A set breakpoint in the function (rust-side) that should be invoked, will be disabled, when the debugger starts.
Although the breakpoint in the main() function still works.

Maybe you have an idea what causes this issue?

Many thanks

**Reply by FabianLars** [reply]
> Maybe you have an idea what causes this issue?

Not reallyyy, but most of the times, problems with invoke should get logged in the devtools console (right click -> inspect). Does it work correctly if used with `tauri dev`? (there shouldn't be a difference 🤔)

**Reply by ReinhardKrisam** [reply]
Hmmm, maybe the restart of vscode has done the miracle...now it works.

The breakpoints in main() as well as in the invoked function are working now.

I can continue developing with tauri 😊👍

Thanks a lot for your help.

**Comment by kieuminhcanh** [comment]
@ReinhardKrisam, @FabianLars , @damien, @zuckschwerdt  You guys can do via 2 steps

1. Build **Dev** App and Let app run for the first time and then close.
```sh
yarn tauri dev
```

2. Edit your **.vscode\launch.json**
```json
{  
  "version": "0.2.0",
  "configurations": [
    {
      "name": "Front-end App Debug",
      "type": "msedge",
      "port": 9222,
      "request": "launch",
      "runtimeExecutable": "${workspaceFolder}/src-tauri/target/debug/app.exe", // Remember to edit the right path to your app.
      "useWebView": true,
      "webRoot": "${workspaceFolder}",
      "preLaunchTask": "ui:dev"
    },   
  ],
}
```

**.vscode\tasks.json**
```json
{
  "version": "2.0.0",
  "tasks": [
    {
      "label": "ui:dev",
      "type": "shell",
      "isBackground": true,
      "command": "yarn",
      "args": [
        "dev"
      ],
    },
  ]
}
```
Open **vscode** > Select **Front-end App Debug** > F5 for start debug.

**Reply by Ricord** [reply]
I see the tasks.json for yarn, may I ask how to apply to npm?

**Reply by FabianLars** [reply]
Pretty sure that you just have to replace `yarn` with `npm`. vscode should handle the rest.

**Reply by pgenfer** [reply]
@kieuminhcanh   If I use your launch configuration to debug the WebView process, I get the following error:

> Unable to launch browser: "Could not connect to debug target at http://localhost:9222. Could not find any debuggable target"

The Tauri application starts fine and works without problems, however it seems the VS Code debugger cannot connect. Is there maybe any configuration I'm missing?

Thanks a lot for your help!

**Reply by kieuminhcanh** [reply]
@pgenfer 
Those settings just work with windows OS only. Could you share your launch.config and task.config?

**Reply by pgenfer** [reply]
@kieuminhcanh 
Ah, thanks for the info, that was my mistake. 
Tried it with macOS (using `type: 'chrome'` instead of `msedge`).

I checked your solution on Windows and there it works like a charm!

Just found this [discussion](https://github.com/tauri-apps/tauri-docs/issues/842), I guess frontend debugging via VSCode for macOS/Linux is not yet supported...

Anyway, thanks a lot for your support!
Greetings,
Patric
